### PR TITLE
Add example for ALPN usage.

### LINKED ISSUE #807: ALPN needs documenting
Currently, there is no documentation on how one would go about implementing ALPN using SslStream. This should be fixed as ALPN is becoming more widely used.

---
#### Document details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 4f3f9002-51ab-801e-988e-0ca3008c69f8
* Version Independent ID: 1b3aab2f-4916-6756-74d3-3aa32a9704a5
* Content: [SslApplicationProtocol Struct (System.Net.Security)](https://docs.microsoft.com/en-gb/dotnet/api/system.net.security.sslapplicationprotocol?view=netcore-2.1)
* Content Source: [xml/System.Net.Security/SslApplicationProtocol.xml](https://github.com/dotnet/dotnet-api-docs/blob/master/xml/System.Net.Security/SslApplicationProtocol.xml)
* Product: **.net**
* GitHub Login: @dotnet-bot
* Microsoft Alias: **dotnetcontent**

### PR DESCRIPTION
**Pull request by rzikm**
Fixes #807.

### CODE REVIEW

**Comment by rzikm** [comment]
bump @dotnet/ncl, please take a look.

File: snippets/csharp/System.Net.Security/ALPN/alpn.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+using System.Net;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading.Tasks;
+
+async Task RunServer()
+{
+    using X509Certificate2 serverCertificate = new X509Certificate2("contoso.com.pfx", "testcertificate");
```
Comment (liveans) [inline snippets/csharp/System.Net.Security/ALPN/alpn.cs:11]:
```suggestion
    using X509Certificate2 serverCertificate = new("contoso.com.pfx", "testcertificate");

```

File: snippets/csharp/System.Net.Security/ALPN/alpn.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    TcpListener listener = new TcpListener(IPAddress.Loopback, 5004);
+    listener.Start();
+    using TcpClient server = await listener.AcceptTcpClientAsync();
+    listener.Stop();
```
Comment (liveans) [inline snippets/csharp/System.Net.Security/ALPN/alpn.cs:16]:
```suggestion
    var listener = new TcpListener(IPAddress.Loopback, 5004);
    listener.Start();
    using var server = await listener.AcceptTcpClientAsync();
    listener.Stop();

```

File: snippets/csharp/System.Net.Security/ALPN/alpn.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    await Server(server.GetStream(), serverCertificate);
+}
+
+async Task RunClient()
+{
+    using TcpClient client = new TcpClient();
```
Comment (liveans) [inline snippets/csharp/System.Net.Security/ALPN/alpn.cs:23]:
```suggestion
    using var client = new TcpClient();

```

File: snippets/csharp/System.Net.Security/ALPN/alpn.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    await client.ConnectAsync(new IPEndPoint(IPAddress.Loopback, 5004));
+
+    await Client(client.GetStream(), "testserver.contoso.com");
+}
+
+//<snippet>
+async Task Server(NetworkStream stream, X509Certificate2 serverCertificate)
+{
+    using SslStream server = new SslStream(stream);
```
Comment (liveans) [inline snippets/csharp/System.Net.Security/ALPN/alpn.cs:32]:
```suggestion
    using var server = new SslStream(stream);

```

File: snippets/csharp/System.Net.Security/ALPN/alpn.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    await server.AuthenticateAsServerAsync(new SslServerAuthenticationOptions
+    {
+        ServerCertificate = serverCertificate,
+        ApplicationProtocols = new List<SslApplicationProtocol>
+        {
+            new SslApplicationProtocol("protocol1"),
+            new SslApplicationProtocol("protocol2"),
+        },
```
Comment (liveans) [inline snippets/csharp/System.Net.Security/ALPN/alpn.cs:40]:
```suggestion
        ApplicationProtocols = new()
        {
            new("protocol1"),
            new("protocol2"),
        },

```

File: snippets/csharp/System.Net.Security/ALPN/alpn.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    });
+
+    string protocol = Encoding.ASCII.GetString(server.NegotiatedApplicationProtocol.Protocol.Span);
+    System.Console.WriteLine($"Server - negotiated protocol: {protocol}");
+}
+
+async Task Client(NetworkStream stream, string hostName)
+{
+    using SslStream client = new SslStream(stream);
```
Comment (liveans) [inline snippets/csharp/System.Net.Security/ALPN/alpn.cs:49]:
```suggestion
    using var client = new SslStream(stream);

```

File: snippets/csharp/System.Net.Security/ALPN/alpn.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    await client.AuthenticateAsClientAsync(new SslClientAuthenticationOptions
+    {
+        // the host name must match the name on the certificate used on the server side
+        TargetHost = hostName,
+        ApplicationProtocols = new List<SslApplicationProtocol>
+        {
+            new SslApplicationProtocol("protocol2"),
+            new SslApplicationProtocol("protocol3")
+        },
```
Comment (liveans) [inline snippets/csharp/System.Net.Security/ALPN/alpn.cs:58]:
```suggestion
        ApplicationProtocols = new()
        {
            new("protocol2"),
            new("protocol3")
        },

```

**Review by liveans — APPROVED** [review]
Up to you to commit suggested changes but LGTM anyway.

File: snippets/csharp/System.Net.Security/ALPN/alpn.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+{
+    using X509Certificate2 serverCertificate = new("contoso.com.pfx", "testcertificate");
+
+
+    var listener = new TcpListener(IPAddress.Loopback, 5004);
+    listener.Start();
+    using var server = await listener.AcceptTcpClientAsync();
+    listener.Stop();
+
+
```
Comment (antonfirsov) [inline snippets/csharp/System.Net.Security/ALPN/alpn.cs:19]:
```suggestion

```

File: snippets/csharp/System.Net.Security/ALPN/alpn.cs [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    await server.AuthenticateAsServerAsync(new SslServerAuthenticationOptions
+    {
+        ServerCertificate = serverCertificate,
+        ApplicationProtocols = new()
+        {
+            new("protocol1"),
+            new("protocol2"),
+        },
+
```
Comment (antonfirsov) [inline snippets/csharp/System.Net.Security/ALPN/alpn.cs:45]:
```suggestion
        }
```

File: snippets/csharp/System.Net.Security/ALPN/alpn.cs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    await client.AuthenticateAsClientAsync(new SslClientAuthenticationOptions
+    {
+        // the host name must match the name on the certificate used on the server side
+        TargetHost = hostName,
+        ApplicationProtocols = new()
+        {
+            new("protocol2"),
+            new("protocol3")
+        },
+
```
Comment (antonfirsov) [inline snippets/csharp/System.Net.Security/ALPN/alpn.cs:65]:
```suggestion
        }
```

**Review by antonfirsov — APPROVED** [review]
nits

**Pull request merged**